include/boost/compute/algorithm/detail/scan_on_cpu.hpp

```diff
@@ -36,15 +38,32 @@ inline OutputIterator scan_on_cpu(InputIterator first,
 
     typedef typename
         std::iterator_traits<InputIterator>::value_type input_type;
+    typedef typename
+        std::iterator_traits<OutputIterator>::value_type output_type;
 
     const context &context = queue.get_context();
 
     // create scan kernel
     meta_kernel k("scan_on_cpu");
-    k.add_arg<ulong_>("n");
+
+    // Arguments
+    size_t n_arg = k.add_arg<ulong_>("n");
+    size_t init_arg = k.add_arg<output_type>("init");
+
+    if(!exclusive){
+        k <<
+            k.decl<const ulong>("start_idx") << " = 1;\n" <<
+            k.decl<output_type>("sum") << " = " << first[0] << ";\n" <<
```

Comment:
It's usually better to not directly embed user data inside the kernel source, this leads to kernel recompilation for each different input range. Rather it's better to either pass it as a kernel argument or look it up in the input range inside the kernel.